pkg/controller/iam/managed.go

```diff
@@ -0,0 +1,261 @@
+	service, err := iamv1.NewService(ctx, option.WithCredentialsJSON(credentials))
+	if err != nil {
+		return nil, err
+	}
+	return iamv1.NewProjectsService(service).ServiceAccounts, nil
+}
+
+type connecter struct {
+	client client.Client
+	newSAS func(ctx context.Context, creds []byte) (*iamv1.ProjectsServiceAccountsService, error)
+}
+
+// Connect sets up iam client using credentials from the provider
+func (c *connecter) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	cr, ok := mg.(*v1beta1.ServiceAccount)
+	if !ok {
+		return nil, errors.New(errNotServiceAccount)
+	}
+
+	p := &gcpv1alpha3.Provider{}
+	if err := c.client.Get(ctx, meta.NamespacedNameOf(cr.Spec.ProviderReference), p); err != nil {
+		return nil, errors.Wrap(err, errGetProvider)
+	}
+
+	if p.GetCredentialsSecretReference() == nil {
+		return nil, errors.New(errProviderSecretRef)
+	}
+
+	s := &corev1.Secret{}
+	n := types.NamespacedName{Namespace: p.Spec.CredentialsSecretRef.Namespace, Name: p.Spec.CredentialsSecretRef.Name}
+	if err := c.client.Get(ctx, n, s); err != nil {
+		return nil, errors.Wrap(err, errGetProviderSecret)
+	}
+
+	saAPI, err := c.newSAS(ctx, s.Data[p.Spec.CredentialsSecretRef.Key])
+	rrn := NewRelativeResourceNamer(p.Spec.ProjectID)
+	return &external{serviceAccounts: saAPI, rrn: rrn}, errors.Wrap(err, errNewClient)
+}
+
+type external struct {
+	serviceAccounts *iamv1.ProjectsServiceAccountsService
+	rrn             RelativeResourceNamer
+}
+
+func (e *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) { // nolint:gocyclo
+	cr, ok := mg.(*v1beta1.ServiceAccount)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotServiceAccount)
+	}
+
+	getRequest := e.serviceAccounts.Get(e.rrn.ResourceName(cr))
+	fromProvider, err := getRequest.Context(ctx).Do()
+	if err != nil {
+		if gcp.IsErrorNotFound(err) {
+			return managed.ExternalObservation{ResourceExists: false}, nil
+		}
+		return managed.ExternalObservation{}, errors.Wrap(err, errGet)
+	}
+
+	populateCRFromProvider(cr, fromProvider)
+	return managed.ExternalObservation{
+		ResourceExists:    true,
+		ResourceUpToDate:  isUpToDate(&cr.Spec.ForProvider, fromProvider),
+		ConnectionDetails: managed.ConnectionDetails{},
+	}, nil
+}
+
+// https://cloud.google.com/iam/docs/reference/rest/v1/projects.serviceAccounts/create
+// Note that the metadata.Name from the Kubernetes custom resource is used as the AccountID parameter
+// All other API methods use the external-name annotation
+// (set via the RelativeResourceNameAsExternalName Initializer)
+func (e *external) Create(ctx context.Context, mg resource.Managed) (managed.ExternalCreation, error) {
+	cr, ok := mg.(*v1beta1.ServiceAccount)
+	if !ok {
+		return managed.ExternalCreation{}, errors.New(errNotServiceAccount)
+	}
+
+	csar := &iamv1.CreateServiceAccountRequest{
+		AccountId: cr.GetName(),
```

Comment:
```suggestion
		AccountId: meta.GetExternalName(cr),
```

We wanted to use the `crossplane.io/external-name` annotation here, right? The suggested function will return its value (which will typically be `cr.GetName()` unless the user specified the annotation).

Reply:
Agreed that this could be useful in `crossplane-runtime`. There is similar logic already present in crossplane and it would be good to have a standard across the board: https://github.com/crossplane/crossplane/tree/master/pkg/stacks/truncate

Reply:
derp!

Reply:
derp was re using the wrong method, Muvaf and Dan's comments weren't showing at the time :)

re Muvaf's suggestion, I am still leaning toward not transforming the value, pushing the problem upward. Could we chat about this today? FWIW I have actually implemented the Initializer he is suggesting 1.5x now amongst all these PR loops, so I can write it again with my eyes closed XD

Reply:
@kasey @muvaf @hasheddan Let's chat about this today - I'm not a fan of special casing this logic, even if we do perceive this kind of resource to be more likely than usual to have a name overflow. I want to get on the same page so we aren't giving @kasey conflicting preferences. :)

Reply:
The composition case that I was worried is solved on the other side; we won't randomize the cluster scoped instance so the only addition to the managed resource will be 6 chars instead of 12.

Reply:
Alright I'm going to cal this settled for now, looking forward to figuring this out for the general case. Thanks for all the helpful discussion on this issue!